### PR TITLE
Add No namespace option

### PR DESCRIPTION
**Pull request by michael-pomelo**
This fixes the failing tests that were present in #117

### CODE REVIEW

**Comment by thesayyn** [comment]
thanks for taking this on. tests are still failing.

**Comment by michaelcacciatore** [comment]
They are, but they are different than last time! Will try to fix them today.

Progress!!

**Comment by michael-pomelo** [comment]
Hey @thesayyn it unfortunately looks like main is broken from this commit: https://github.com/thesayyn/protoc-gen-ts/commit/1922704434b95ed69fe9e20594f959a5b6fdab95 (it was committed directly to main so it skipped CI checks).

It adds a private static field to each class that uses a private identifier (`#`). This is not valid Typescript (in the version defined in the repository; more on this later). Running `yarn test` in the `examples/pure` directory gives us this error from the typescript compiler:

```
src/test.ts:12:5 - error TS18019: 'static' modifier cannot be used with a private identifier.
12     static #one_of_decls = [];
```

I checked out out the previous commit (https://github.com/thesayyn/protoc-gen-ts/commit/beee8e9b1c5b5b6103097912af6a133b94204cb6) and the tests run fine just to confirm.

Since I am not familiar with the intention or context behind that commit, I don't think I feel comfortable fixing it on my own.

However, Typescript 4.3.x and above DO support this syntax (release notes: https://www.typescriptlang.org/docs/handbook/release-notes/typescript-4-3.html#static-index-signatures).  Bumping the TS version across the repository to 4.3.5 should fix this problem. I did this in my latest commit but please let me know if you don't think it is the best solution.

**Comment by thesayyn** [comment]
Great! also, I believe it would make more sense to rename the option to `no_toplevel_namespace`.  Also, tests should cover cases such as nested messages, and multiple files importing deeply nested symbols without bumping into the no top-level namespace issue.

```
#test.proto
package test.xd

message T {
  message X {}
}
```

```
import "test.proto"
message Y {
 test.xd.T.X t = 1;
}
```
this case should work with/without this option present.

**Comment by thesayyn** [comment]
Let me take a look at that.

**Comment by thesayyn** [comment]
Hmm, forcing users to one minor version of typescript is not something we want to do. I'll send a PR to make it non-static

File: examples/pure/package.json [hunk 1]
```diff
@@ -11,6 +11,6 @@
     },
     "devDependencies": {
         "protoc-gen-ts": "file:../../bazel-bin/package",
-        "typescript": "^4.2.4"
+        "typescript": "^4.3.5"
```
Comment (thesayyn) [inline examples/pure/package.json:14]: See #133

File: src/option.ts [hunk 2]
```diff
@@ -1,6 +1,7 @@
 export interface Options {
   unary_rpc_promise: boolean;
   grpc_package: string;
+  no_namespace: boolean
```
Comment (thesayyn) [inline src/option.ts:4]: Lets rename this to `no_toplevel_namespace`

**Review by thesayyn — CHANGES_REQUESTED** [review]
Also, test cases are not covering every case.  we should introduce a few protos that import each other with deeply nested enums, rpcs etc.

**Comment by michael-pomelo** [comment]
Reverted the TS upgrade.  Thanks for hopping on the static issue so quickly!

I can definitely write some more tests

**Comment by michael-pomelo** [comment]
I also want to point out that the goal is not to remove only the top level namespace--it is to remove all of them.

I'm not sure what the intention behind the original PR was, but my use case is that `babel-typescript` is incompatible with typescript namespaces, so we can't have them anywhere.

**Comment by michael-pomelo** [comment]
Oh I definitely misunderstood the intention then, my sincerest apologies.

This PR removes all namespaces--even from nested protos.

I can make a 2nd option for removing the top level namespace only (and add more tests if you like).

**Comment by thesayyn** [comment]
> This PR removes all namespaces--even from nested protos.

That's something I would love to see. but beware that's not an easy thing to do given that you have to change the way functions inside `type.ts` works. submessages has to change their class names depending on this option. so lots of internal functions have to behave differently.

**Comment by thesayyn** [comment]
I could provide the necessary information to point you in the right direction.

**Comment by michael-pomelo** [comment]
Just added a more complicated test with nested/shared protos to show that this removes both the top level namespace and any nested ones.

Let me know if there is some context I am missing here!  Also if this seems like it fits the requirements, let me know if you want me to make a separate option/PR for no top level namespaces.

**Comment by thesayyn** [comment]
I just had a glance at it but it still lacks support for nested message types.

See: "test/_/issues/complex.proto" as an example.

**Comment by thesayyn** [comment]
also, you could build the plugin and run against protos inside `test/_` folder. I am sure there will be lots of generation errors.

**Comment by michael-pomelo** [comment]
Appreciate the guidance @thesayyn 

I just pushed a new commit which addresses nested enums (and uses the complex.proto as a dependency of no_namespace.proto).

**Comment by thesayyn** [comment]
> I just pushed a new commit which addresses nested enums (and uses the complex.proto as a dependency of no_namespace.proto).

Also, messages can have sub-messages nested inside them. they have to be handled the same way enums are handled.

File: src/option.ts (hunk 2 above)
Comment (michael-pomelo) [reply src/option.ts:4]: Going to resolve this since the scope of this diff changed to all namespaces.

**Comment by michael-pomelo** [comment]
> Also, messages can have sub-messages nested inside them. they have to be handled the same way enums are handled.

Like `SchedulingContext.Batch` in `complex.proto`? Or something else?

**Comment by thesayyn** [comment]
Exactly.

**Comment by michael-pomelo** [comment]
Modified the test to ensure nested messages work!

**Comment by thesayyn** [comment]
> Modified the test to ensure nested messages work!

Hey. Sorry, I was away because of relocating.  Great work so far!!

I'll have another pass on the changes. We'll definitely land this after we get it into good shape.

File: src/option.ts [hunk 3]
```diff
@@ -27,5 +29,7 @@ export function parse(raw?: string): Options {
     }
     options[k] = value
   }
+  // @ts-ignore
+  global.no_namespace = options.no_namespace;
```
Comment (thesayyn) [inline src/option.ts:33]: Uh oh, this looks nasty.  is there any way we could avoid doing this? may be i could help if you can explain the reason why?

File: src/type.ts [hunk 4]
```diff
@@ -1,9 +1,17 @@
 import * as descriptor from "./compiler/descriptor.js";
```
Comment (thesayyn) [inline src/type.ts:1]: type registry shouldn't be getting this much involved in what options have been passed. I know that when you strip out the root package name it is really hard to find out what part of type_name is the package but I still feel like we could be move clever about this.

perhaps we could keep track of things twice with and without their package names to locate them quicker?  for instance, leading dot is a good clue to find out whether a symbol comes from the same file or from a dependency.  I'll think about this one.

File: src/index.ts [hunk 5]
```diff
@@ -39,6 +39,8 @@ const response = new plugin.CodeGeneratorResponse({
 
 const options = op.parse(request.parameter);
 
+type.initialize(options);
```
Comment (thesayyn) [inline src/index.ts:42]: this is a red flag. see my comments on type.ts

**Review by thesayyn — CHANGES_REQUESTED** [review]

File: src/option.ts (hunk 3 above)
Comment (michael-pomelo) [reply src/option.ts:33]: This was just me fiddling/debugging and should not have made it into the commit.

That's why this value isn't read anywhere

Comment (michael-pomelo) [reply src/option.ts:33]: I'll remove it in the next round of changes

File: src/type.ts (hunk 4 above)
Comment (michael-pomelo) [reply src/type.ts:1]: Totally understand what you're saying; how we pass the options feels sort of like an escape hatch.

But at the end of the day, both `getTypeReference` and `getTypeReferenceExpr` need to know if we are using namespaces or not. Let me know what your thoughts are!

File: src/option.ts (hunk 3 above)
Comment (thesayyn) [reply src/option.ts:33]: Ah, all right. i was scared for a second there 🤪

File: src/type.ts (hunk 4 above)
Comment (thesayyn) [reply src/type.ts:1]: yep. I see your point. perhaps it is time to introduce something like compilation context using context pattern but I'll follow up with another PR.

Comment (thesayyn) [reply src/type.ts:1]: also, packageName is kind of confusing perhaps we should rename it to parentName.

File: src/option.ts (hunk 3 above)
Comment (michael-pomelo) [reply src/option.ts]: Resolving since this is removed in latest commit

File: src/index.ts (hunk 5 above)
Comment (michael-pomelo) [reply src/index.ts:42]: Resolved as per the other thread (adding compilation context in a future PR)

File: src/type.ts (hunk 4 above)
Comment (michael-pomelo) [reply src/type.ts:1]: Changed!

**Pull request merged**